drivers/can/can_socketcan.c

```diff
@@ -0,0 +1,161 @@
+	if (!pkt) {
+		LOG_ERR("Failed to obtain net_pkt");
+		return;
+	}
+
+	if (net_pkt_write(pkt, frame, sizeof(struct zcan_frame))) {
+		LOG_ERR("Failed to append RX data");
+		net_pkt_unref(pkt);
+		return;
+	}
+
+	ret = net_recv_data(ctx->iface, pkt);
+	if (ret < 0) {
+		LOG_DBG("net_recv_data failed [%d]", ret);
+		net_pkt_unref(pkt);
+	}
+}
+
+static int socket_can_setsockopt(const struct device *dev, void *obj,
+					int level, int optname,
+					const void *optval, socklen_t optlen)
+{
+	const struct socket_can_config *cfg = dev->config;
+	struct socket_can_context *socket_context = dev->data;
+	struct net_context *ctx = obj;
+	int ret;
+
+	if (level != SOL_CAN_RAW && optname != CAN_RAW_FILTER) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	__ASSERT_NO_MSG(optlen == sizeof(struct zcan_filter));
+
+	ret = can_add_rx_filter(cfg->can_dev, socket_can_recv, socket_context, optval);
+	if (ret == -ENOSPC) {
+		errno = ENOSPC;
+		return -1;
+	}
+
+	net_context_set_filter_id(ctx, ret);
+
+	return 0;
+}
+
+static void socket_can_close(const struct device *dev, int filter_id)
+{
+	const struct socket_can_config *cfg = dev->config;
+
+	can_remove_rx_filter(cfg->can_dev, filter_id);
+}
+
+static void socket_can_send_tx_callback(int error, void *arg)
+{
+	if (error != 0) {
+		LOG_DBG("socket CAN TX error [%d]", error);
+	}
+}
+
+static int socket_can_send(const struct device *dev,
+				  struct net_pkt *pkt)
+{
+	const struct socket_can_config *cfg = dev->config;
+	int ret;
+
+	if (net_pkt_family(pkt) != AF_CAN) {
+		return -EPFNOSUPPORT;
+	}
+
+	ret = can_send(cfg->can_dev,
+		       (struct zcan_frame *)pkt->frags->data,
+		       SEND_TIMEOUT, socket_can_send_tx_callback, NULL);
+
+	if (ret == 0) {
+		net_pkt_unref(pkt);
+	} else {
+		LOG_DBG("Cannot send socket CAN msg (%d)", ret);
+	}
+
+	/* If something went wrong, then we need to return negative value to
+	 * net_if.c:net_if_tx() so that the net_pkt will get released.
+	 */
+	return ret;
+}
+
+static void socket_can_iface_init(struct net_if *iface)
+{
+	const struct device *dev = net_if_get_device(iface);
+	struct socket_can_context *socket_context = dev->data;
+
+	socket_context->iface = iface;
+
+	LOG_DBG("Init CAN interface %p dev %p", iface, dev);
+}
+
+static int socket_can_init(const struct device *dev)
+{
+	const struct socket_can_config *cfg = dev->config;
+
+	ARG_UNUSED(cfg);
+
+	LOG_DBG("Init socket CAN device %p (%s) for dev %p (%s)",
+		dev, dev->name, cfg->can_dev, cfg->can_dev->name);
+
+	return 0;
+}
+
+static struct canbus_api socket_can_api = {
+	.iface_api.init = socket_can_iface_init,
+	.send = socket_can_send,
+	.close = socket_can_close,
+	.setsockopt = socket_can_setsockopt,
+};
+
+static struct socket_can_context socket_can_ctx;
+
+static const struct socket_can_config socket_can_cfg = {
+	.can_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_canbus))
+};
+
+NET_DEVICE_INIT(socket_can, "SOCKET_CAN", socket_can_init,
+		NULL, &socket_can_ctx, &socket_can_cfg,
+		CONFIG_CAN_INIT_PRIORITY,
```

Comment:
Won't having this at the same initialisation priority as the CAN controller drivers cause a race condition?

Reply:
Good point. Was always like this.
Reused the duplicated init priority instead